File: plugins/vts-driver/cerise-plugin.js
/**
 * Cerise Plugin SDK for Node.js
 *
 * Use this SDK to create plugins that communicate with Cerise Core.
 */

const readline = require('readline');

/**
 * @typedef {Object} AbilityContext
 * @property {string} user_id
 * @property {string} session_id
 * @property {string[]} permissions
 */

/**
 * @typedef {Object} AbilityResult
 * @property {boolean} success
 * @property {any} data
 * @property {string|null} error
 * @property {string|null} emotion_hint
 */

/**
 * Base class for Cerise plugins
 * @abstract
 */
class BasePlugin {
  constructor() {
    /** @type {Object} */
    this.config = {};
    /** @type {string[]} */
    this.permissions = [];
  }

  /**
   * Return list of abilities provided by this plugin
   * @abstract
   * @returns {Array<{name: string, description: string, parameters: Object}>}
   */
  getAbilities() {
    throw new Error('getAbilities() must be implemented');
  }

  /**
   * Execute an ability
   * @abstract
   * @param {string} ability - Ability name
   * @param {Object} params - Parameters
   * @param {AbilityContext} context - Execution context
   * @returns {Promise<AbilityResult>}
   */
  async execute(ability, params, context) {
    throw new Error('execute() must be implemented');
  }

  /**
   * Called when plugin is initialized
   * @param {Object} config - Configuration
   * @returns {Promise<boolean>}
   */
  async onInitialize(config) {
    this.config = config;
    return true;
  }

  /**
   * Called when plugin is shutting down
   * @returns {Promise<void>}
   */
  async onShutdown() {}
}

/**
 * Runs a plugin and handles JSON-RPC communication
 */
class PluginRunner {
  /**
   * @param {BasePlugin} plugin
   */
  constructor(plugin) {
    this.plugin = plugin;
    this._running = false;
    this._rl = null;
  }

  /**
   * Main loop: read from stdin, write to stdout
   * @returns {Promise<void>}
   */
  async run() {
    this._running = true;

    this._rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
      terminal: false,
    });

    for await (const line of this._rl) {
      if (!this._running) break;

      try {
        const request = JSON.parse(line.trim());
        const response = await this._handleRequest(request);

        if (response) {
          console.log(JSON.stringify(response));
        }
      } catch (e) {
        if (e instanceof SyntaxError) {
          const errorResponse = {
            jsonrpc: '2.0',
            error: { code: -32700, message: `Parse error: ${e.message}` },
            id: null,
          };
          console.log(JSON.stringify(errorResponse));
        } else {
          const errorResponse = {
            jsonrpc: '2.0',
            error: { code: -32603, message: `Internal error: ${e.message}` },
            id: null,
          };
          console.log(JSON.stringify(errorResponse));
        }
      }
    }
  }

  /**
   * Handle a JSON-RPC request
   * @param {Object} request
   * @returns {Promise<Object|null>}
   */
  async _handleRequest(request) {
    const method = request.method || '';
    const params = request.params || {};
    const reqId = request.id;

    const isNotification = reqId === undefined || reqId === null;

    try {
      const result = await this._dispatch(method, params);
      if (isNotification) return null;
      return {
        jsonrpc: '2.0',
        result,
        id: reqId,
      };
    } catch (e) {
      if (isNotification) return null;
      return {
        jsonrpc: '2.0',
        error: { code: -32603, message: e.message },
        id: reqId,
      };
    }
  }

  /**
   * Dispatch to appropriate handler
   * @param {string} method
   * @param {Object} params
   * @returns {Promise<any>}
   */
  async _dispatch(method, params) {
    switch (method) {
      case 'initialize': {
        const config = params.config || {};
        this.plugin.permissions = params.permissions || [];
        const success = await this.plugin.onInitialize(config);
        const abilities = this.plugin.getAbilities();
        return {
          success,
          abilities,
          skills: abilities,
          tools: abilities,
        };
      }

      case 'execute': {
        const ability = params.ability || params.skill || params.tool || params.name || '';
        const execParams = params.params ?? params.arguments ?? {};
        const contextData = params.context || {};

        const context = {
          user_id: contextData.user_id || '',
          session_id: contextData.session_id || '',
          permissions: contextData.permissions || [],
        };

        const result = await this.plugin.execute(ability, execParams, context);
        return {
          success: result.success,
          data: result.data,
          error: result.error,
          emotion_hint: result.emotion_hint,
        };
      }

      case 'health':
        return { healthy: true };

      case 'shutdown':
        await this.plugin.onShutdown();
        this._running = false;
        if (this._rl) this._rl.close();
        return { success: true };

      default:
        throw new Error(`Unknown method: ${method}`);
    }
  }

  /**
   * Send an event notification to Core
   * @param {string} eventType
   * @param {Object} data
   */
  sendEvent(eventType, data) {
    const notification = {
      jsonrpc: '2.0',
      method: 'event',
      params: { type: eventType, data },
    };
    console.log(JSON.stringify(notification));
  }

  /**
   * Send a log message to Core
   * @param {string} level
   * @param {string} message
   */
  log(level, message) {
    const notification = {
      jsonrpc: '2.0',
      method: 'log',
      params: { level, message },
    };
    console.log(JSON.stringify(notification));
  }
}

/**
 * Run a plugin (blocking)
 * @param {BasePlugin} plugin
 */
function runPlugin(plugin) {
  const runner = new PluginRunner(plugin);
  runner.run().catch((e) => {
    console.error('Plugin error:', e);
    process.exit(1);
  });
}

module.exports = {
  BasePlugin,
  PluginRunner,
  runPlugin,
};

File: sdk/go/cerise_plugin.go
// Cerise Plugin SDK for Go
//
// This SDK implements the newline-delimited JSON-RPC protocol used by Cerise Core.
// A plugin runs as a subprocess (stdio transport) and responds to:
// - initialize
// - execute
// - health
// - shutdown
package ceriseplugin

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
)

type AbilityContext struct {
	UserID      string   `json:"user_id"`
	SessionID   string   `json:"session_id"`
	Permissions []string `json:"permissions"`
}

type AbilityResult struct {
	Success     bool        `json:"success"`
	Data        any         `json:"data,omitempty"`
	Error       *string     `json:"error,omitempty"`
	EmotionHint *string     `json:"emotion_hint,omitempty"`
}

type Ability struct {
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	Parameters  map[string]any `json:"parameters,omitempty"`
}

type Plugin interface {
	GetAbilities() []Ability
	OnInitialize(config map[string]any) error
	Execute(ability string, params map[string]any, ctx AbilityContext) (AbilityResult, error)
	OnShutdown() error
}

type jsonrpcRequest struct {
	JSONRPC string         `json:"jsonrpc"`
	Method  string         `json:"method"`
	Params  map[string]any `json:"params,omitempty"`
	ID      any            `json:"id,omitempty"`
}

type jsonrpcResponse struct {
	JSONRPC string         `json:"jsonrpc"`
	Result  any            `json:"result,omitempty"`
	Error   *jsonrpcError  `json:"error,omitempty"`
	ID      any            `json:"id,omitempty"`
}

type jsonrpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

func Run(plugin Plugin) error {
	in := bufio.NewScanner(os.Stdin)
	// Allow larger payloads than the default 64K.
	in.Buffer(make([]byte, 0, 1024*1024), 8*1024*1024)

	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	running := true
	for running && in.Scan() {
		line := in.Bytes()
		if len(line) == 0 {
			continue
		}

		var req jsonrpcRequest
		if err := json.Unmarshal(line, &req); err != nil {
			resp := jsonrpcResponse{
				JSONRPC: "2.0",
				Error:   &jsonrpcError{Code: -32700, Message: fmt.Sprintf("Parse error: %v", err)},
				ID:      nil,
			}
			_ = writeJSON(out, resp)
			continue
		}

		// Notifications have no id; ignore response.
		isNotification := req.ID == nil

		result, rpcErr, shouldStop := dispatch(plugin, req.Method, req.Params)
		if shouldStop {
			running = false
		}

		if isNotification {
			continue
		}

		resp := jsonrpcResponse{JSONRPC: "2.0", ID: req.ID}
		if rpcErr != nil {
			resp.Error = rpcErr
		} else {
			resp.Result = result
		}

		_ = writeJSON(out, resp)
	}

	return in.Err()
}

func dispatch(plugin Plugin, method string, params map[string]any) (any, *jsonrpcError, bool) {
	switch method {
	case "initialize":
		cfg, _ := params["config"].(map[string]any)
		if cfg == nil {
			cfg = map[string]any{}
		}
		if err := plugin.OnInitialize(cfg); err != nil {
			return map[string]any{"success": false, "error": err.Error()}, nil, false
		}
		abilities := plugin.GetAbilities()
		return map[string]any{
			"success":   true,
			"abilities": abilities,
			"skills":    abilities,
			"tools":     abilities,
		}, nil, false

	case "execute":
		ability, _ := params["ability"].(string)
		if ability == "" {
			ability, _ = params["skill"].(string)
		}
		if ability == "" {
			ability, _ = params["tool"].(string)
		}
		if ability == "" {
			ability, _ = params["name"].(string)
		}

		execParams, _ := params["params"].(map[string]any)
		if execParams == nil {
			execParams, _ = params["arguments"].(map[string]any)
		}
		if execParams == nil {
			execParams = map[string]any{}
		}

		ctxRaw, _ := params["context"].(map[string]any)
		ctx := AbilityContext{}
		if ctxRaw != nil {
			if v, ok := ctxRaw["user_id"].(string); ok {
				ctx.UserID = v
			}
			if v, ok := ctxRaw["session_id"].(string); ok {
				ctx.SessionID = v
			}
			if v, ok := ctxRaw["permissions"].([]any); ok {
				perms := make([]string, 0, len(v))
				for _, p := range v {
					if s, ok := p.(string); ok {
						perms = append(perms, s)
					}
				}
				ctx.Permissions = perms
			}
		}

		res, err := plugin.Execute(ability, execParams, ctx)
		if err != nil {
			msg := err.Error()
			res.Success = false
			res.Error = &msg
		}
		return res, nil, false

	case "health":
		return map[string]any{"healthy": true}, nil, false

	case "shutdown":
		_ = plugin.OnShutdown()
		return map[string]any{"success": true}, nil, true

	default:
		return nil, &jsonrpcError{Code: -32601, Message: "Method not found"}, false
	}
}

func writeJSON(w *bufio.Writer, v any) error {
	b, err := json.Marshal(v)
	if err != nil {
		return err
	}
	_, err = w.Write(append(b, '\n'))
	if err != nil {
		return err
	}
	return w.Flush()
}

